include/nbl/video/ILogicalDevice.h

```diff
@@ -539,19 +543,133 @@ class ILogicalDevice : public core::IReferenceCounted
+                resetFences(1u, &fence);
+                cmdbuf->reset(IGPUCommandBuffer::ERF_RELEASE_RESOURCES_BIT);
+                cmdbuf->begin(IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
+                continue;
+            }
+            // some platforms expose non-coherent host-visible GPU memory, so writes need to be flushed explicitly
+            if (m_defaultUploadBuffer.get()->needsManualFlushOrInvalidate())
+            {
+                IDriverMemoryAllocation::MappedMemoryRange flushRange(m_defaultUploadBuffer.get()->getBuffer()->getBoundMemory(), localOffset, subSize);
+                flushMappedMemoryRanges(1u, &flushRange);
+            }
+            // after we make sure writes are in GPU memory (visible to GPU) and not still in a cache, we can copy using the GPU to device-only memory
+            asset::SBufferCopy copy;
+            copy.srcOffset = localOffset;
+            copy.dstOffset = bufferRange.offset + uploadedSize;
+            copy.size = subSize;
+            cmdbuf->copyBuffer(m_defaultUploadBuffer.get()->getBuffer(), bufferRange.buffer.get(), 1u, &copy);
+            // this doesn't actually free the memory, the memory is queued up to be freed only after the GPU fence/event is signalled
+            /*event = */m_defaultUploadBuffer.get()->multi_free(1u, &localOffset, &subSize, core::smart_refctd_ptr<IGPUFence>(fence)); // can queue with a reset but not yet pending fence, just fine
+            uploadedSize += subSize;
+        }
+        //return event.deferred_function; // ????? wtf is deferred_function? why would i have an event here??
+    }
+
+    inline void updateBufferRangeViaStagingBuffer(IGPUFence* fence, IGPUQueue* _queue, const asset::SBufferRange<IGPUBuffer>& bufferRange, const void* data)
+    {
+        core::smart_refctd_ptr<IGPUCommandPool> pool = createCommandPool(_queue->getFamilyIndex(), IGPUCommandPool::ECF_RESET_COMMAND_BUFFER_BIT);
+        core::smart_refctd_ptr<IGPUCommandBuffer> cmdbuf;
+        createCommandBuffers(pool.get(), IGPUCommandBuffer::EL_PRIMARY, 1u, &cmdbuf);
+        assert(cmdbuf);
+        cmdbuf->begin(IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
+        /*auto func = */updateBufferRangeViaStagingBuffer(cmdbuf.get(), fence, _queue, bufferRange, data);
+        cmdbuf->end();
+        IGPUQueue::SSubmitInfo submit;
+        submit.commandBufferCount = 1u;
+        auto* cmdbuf_raw = cmdbuf.get();
+        submit.commandBuffers = &cmdbuf_raw;
+        submit.signalSemaphoreCount = 0u;
+        submit.pSignalSemaphores = nullptr;
+        submit.waitSemaphoreCount = 0u;
+        submit.pWaitSemaphores = nullptr;
+        submit.pWaitDstStageMask = nullptr;
+        _queue->submit(1u, &submit, fence); // <threadsafeSubmit==true> means submit while locking an internal mutex
+        //func.optionalCmdBuffToDrop = std::move(cmdbuf); // ?????? wtf is optionalCmdBuffToDrop?
```

Comment:
in the old design the deferred function was to carry out a `free_addr` of the ranges used in the staging buffer when the GPU is done doing the copy from staging to the destination.

With Vulkan you also need to make sure that the commandbuffer (that the user never has any control over) and its pool (trivial, cmdbuffer has a smartpointer to the pool) does not get dropped before the submit completes, so the deferred function should also be able to take a smartpointer to a submit.